dissect/util/stream.py

```diff
@@ -550,3 +552,95 @@ def _read(self, offset: int, length: int) -> bytes:
             overlay_idx += 1
 
         return b"".join(result)
+
+
+class ZlibStream(AlignedStream):
+    """Create a zlib stream from another file-like object.
+
+    Basically the same as ``gzip.GzipFile`` but for raw zlib streams.
+    Due to the nature of zlib streams, seeking backwards requires resetting the decompression context.
+
+    Args:
+        fh: The source file-like object.
+        size: The size the stream should be.
+    """
+
+    def __init__(self, fh: BinaryIO, size: Optional[int] = None, align: int = STREAM_BUFFER_SIZE, **kwargs):
+        self._fh = fh
+
+        self._zlib = None
+        self._zlib_args = kwargs
+        self._zlib_offset = 0
+        self._zlib_prepend = b""
+        self._zlib_prepend_offset = None
+        self._rewind()
+
+        super().__init__(size, align)
+
+    def _rewind(self) -> None:
+        self._fh.seek(0)
+        self._zlib = zlib.decompressobj(**self._zlib_args)
+        self._zlib_offset = 0
+        self._zlib_prepend = b""
+        self._zlib_prepend_offset = None
+
+    def _seek_zlib(self, offset: int) -> None:
+        if offset < self._zlib_offset:
+            self._rewind()
+
+        while self._zlib_offset < offset:
+            read_size = min(offset - self._zlib_offset, self.align)
+            if self._read_zlib(read_size) == b"":
+                break
+
+    def _read_fh(self, length: int) -> bytes:
+        if self._zlib_prepend_offset is None:
+            return self._fh.read(length)
+
+        if self._zlib_prepend_offset + length <= len(self._zlib_prepend):
+            offset = self._zlib_prepend_offset
+            self._zlib_prepend_offset += length
+            return self._zlib_prepend[offset : self._zlib_prepend_offset]
+        else:
+            offset = self._zlib_prepend_offset
+            self._zlib_prepend_offset = None
+            return self._zlib_prepend[offset:] + self._fh.read(length - len(self._zlib_prepend) + offset)
```

Comment:
I believe this this `else` is not needed.

```suggestion
        offset = self._zlib_prepend_offset
        self._zlib_prepend_offset = None
        return self._zlib_prepend[offset:] + self._fh.read(length - len(self._zlib_prepend) + offset)
```